File: Graph/java/ConnectedComponentsCount.java
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class ConnectedComponentsCount {
  public static int count(Map<String, List<String>> graph) {
    Set<String> visited = new HashSet<>();
    int total = 0;

    for (String node : graph.keySet()) {
      if (explore(graph, node, visited)) {
        total++;
      }
    }

    return total;
  }

  private static boolean explore(Map<String, List<String>> graph, String node, Set<String> visited) {
    if (!visited.add(node)) {
      return false;
    }

    for (String neighbor : graph.getOrDefault(node, List.of())) {
      explore(graph, neighbor, visited);
    }

    return true;
  }
}

File: Graph/python/connectedComponentsCount.py
from .graph_utils import Graph


def connected_components_count(graph: Graph) -> int:
    visited: set[str] = set()
    count = 0

    for node in graph:
        if _explore(graph, node, visited):
            count += 1

    return count


def _explore(graph: Graph, node: str, visited: set[str]) -> bool:
    if node in visited:
        return False
    visited.add(node)

    for neighbor in graph.get(node, []):
        _explore(graph, neighbor, visited)
    return True
